[Shipping Labels] Fix validation of ITN

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Part of WOOMOB-630

> [!WARNING]
> Please don't merge this until the PR #14351 is merged to the `release/22.8` branch and the backmerge PR from the release branch is merged to `trunk`. When this is done, update the base branch of this PR to `trunk`. 

### Description
This PR updates the ITN validation logic, we were missing quite a few cases.

The ITN validation rules are as follows:
- When the destination country is Canada, the ITN is not required.
- When the total value is more than 2500USD (and HS Tariffs not used) then ITN is required.
- When the value of items belonging to the same HS Tariff is more than 2500USD then ITN is required.
- When the destination country is one of `"IR", "KP", "SY", "CU", "SD"` then ITN is required.
- When an ITN is entered, we validate its format.

> [!NOTE]
> The `WooShippingLabelCreationViewModel` still uses the old `ShouldRequireITN` use case, given the size of this PR, I handled this change in a separate PR #14353  

> [!IMPORTANT]
> The rules match what iOS does, the plugin has some differences for now, and we are still waiting for the team confirmation about them p1752754617716169-slack-C05VBLKHHV1 

### Steps to reproduce
Test the steps in this PR https://github.com/woocommerce/woocommerce-ios/pull/15319, just ignore the last two steps, as they are part of the next PR.

### Testing information
Check the above steps.

### The tests that have been performed
The above.

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Introduce ValidateITN use case

hichamboushaba committed:
Add unit tests for ValidateITN use case

hichamboushaba committed:
Update ITN validation rules

- Consider total value cause only when the items don't have an HS tariff number.
- Fix Canada exemption logic
- Include the HS tariff number in the returned error

hichamboushaba committed:
Update unit tests

hichamboushaba committed:
Update the customs form to use the new validation rules

hichamboushaba committed:
Include the HS Tariff number in the error message when requiring an ITN

hichamboushaba committed:
Include only valid HS tariff numbers when checking ITN requirements

hichamboushaba committed:
Fix unit tests

hichamboushaba committed:
Fix detekt issues

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/262">22.8 ❄️</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/customs/WooShippingCustomsFormViewModel.kt:
```diff
@@ -29,78 +32,102 @@ import javax.inject.Inject
     private val navArgs: WooShippingCustomsFormFragmentArgs by savedState.navArgs()
     private val destinationCountryCode = navArgs.destinationCountryCode
 
     private val _viewState = savedState.getStateFlow(
         scope = viewModelScope,
-        initialValue = ViewState()
+        initialValue = initViewState()
```
We were re-initializing the ViewState each time the `ViewModel` was instantiated, which means any use changes are lost in process-death or activity re-creation scenarios. This change ensures we initialize it only once and then restore it from the SavedState if needed.

hichamboushaba committed:
Simplify regex expression

JorgeMucientes committed:
Fix ITN validation by using `value` instead of `totalValue`

The ITN (Internal Transaction Number) validation logic was incorrectly using `totalValue` (quantity * value) instead of the individual item `value` when calculating the total value of customs items. This commit corrects the calculation to use `value` per item, ensuring accurate ITN validation when the total value exceeds $2500.

JorgeMucientes committed:
Use the price value to fill the "Value per unit" input text

JorgeMucientes committed:
Calculate totalValue by multiplying quantity per unit price

JorgeMucientes reviewed (approved):
Good work Hicham! 
Code looks good and most of it worked as expected. However, I had to add a couple fixes to address some issues on how `totalValue` was calculated as well as use the right price per unit value (matching iOS and web behavior). 

Approving and and merging now :shipit:

hichamboushaba commented:
@JorgeMucientes thanks for the review.

Just sharing two remarks here:

- I originally planned to merge this to `trunk` given the big number of changes, I mentioned this in the PR description, but I see it was merged to the `release` branch, I hope the changes are safe given they will be released to production without beta testing.

> I had to add a couple fixes to address some issues on how totalValue was calculated
- Looking at the two commits [43179f4](https://github.com/woocommerce/woocommerce-android/pull/14352/commits/43179f4d543d91cc7364ffa4a937440d599e0fed) and f6f9f37, I don't understand what we did change here, the combined change of the two commits is:

```patch
private val List<CustomsItem>.totalValue: BigDecimal
-        get() = this.sumOf { it.totalValue }
+        get() = this.sumOf { it.value * it.quantity.toBigDecimal() }
```

And knowing that `CustomsItem.totalValue` is already equal to `it.value * it.quantity.toBigDecimal()`, I think we didn't change any logic here, can you explain to me if I'm missing something?

JorgeMucientes commented:
Hey @hichamboushaba, thanks for checking this!

🤦🏼 Sorry, I completely missed the last part of the warning stating I should update the base branch to `trunk`. I just thought I had to wait for the PR https://github.com/woocommerce/woocommerce-android/pull/14351 to be merged first, and then we were good to merge. Luckily, these changes are well-scoped for ITN validation, so I think they are relatively safe to merge as a beta fix. 

Regarding the changes applied: 
> Looking at the two commits [43179f4](https://github.com/woocommerce/woocommerce-android/pull/14352/commits/43179f4d543d91cc7364ffa4a937440d599e0fed) and https://github.com/woocommerce/woocommerce-android/commit/f6f9f370beaf921edf7107ea51803987233f2c17, I don't understand what we did change here, the combined change of the two commits is:

The short answer is that, when attempting to fix a bug, I didn't apply the right fix at first try: [43179f4](https://github.com/woocommerce/woocommerce-android/pull/14352/commits/43179f4d543d91cc7364ffa4a937440d599e0fed) . It was in my second attempt that I fixed the issue in the right place: [1687fc7](https://github.com/woocommerce/woocommerce-android/pull/14352/commits/1687fc73b140c347e325075fa4f7931bb090bebf). 
Then I noticed I had to 

Before this commit [43179f4](https://github.com/woocommerce/woocommerce-android/pull/14352/commits/43179f4d543d91cc7364ffa4a937440d599e0fed), there was a bug in the way we were calculating the total shipment value: 

https://github.com/user-attachments/assets/b06aa76c-74a3-4a63-bf26-5778bfb08f0d

My initial thought was that the issue lay in the way we were calculating the `totalValue` when validating ITN because as you'll see from the screenshot below, `CustomData` field `value` already contained the total value for that shipment, so multiplying it by quantity led to extremely high numbers.

<img width="1639" height="1008" alt="Screenshot 2025-07-25 at 11 42 56" src="https://github.com/user-attachments/assets/7c3f1acb-3da4-42d6-82f7-4050d06b9f6b" />

Then I realised the root cause of the problem was actually in the viewModel: [1687fc7](https://github.com/woocommerce/woocommerce-android/pull/14352/commits/1687fc73b140c347e325075fa4f7931bb090bebf) on how we were assigning the `valuePerUnit` using `shippingTotalValue`. After applying the fix, I noticed that my initial change [43179f4](https://github.com/woocommerce/woocommerce-android/pull/14352/commits/43179f4d543d91cc7364ffa4a937440d599e0fed) was breaking the implementation. However, I completely forgot that: `it.totalValue` was already doing this `it.value * it.quantity.toBigDecimal()` behind the scenes. 
I messed up the cleanliness of the code while fixing the problem, sorry about that. Hope this clarifies what happened.
